File: 2023/07_CamelCards/hands.py

# ======================================================================
# Camel Cards
#   Advent of Code 2023 Day 07 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         h a n d s . p y
# ======================================================================
"Hands for the Advent of Code 2023 Day 07 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
from hand import Hand

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                                  Hands
# ======================================================================


class Hands(object):   # pylint: disable=R0902, R0903, R0205
    "Object for Camel Cards"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.hands = []

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            self._process_text(text)

    def _process_text(self, text):
        "Assign values from text"

        # 0. Preconditions
        assert text is not None and len(text) > 0

        # 1. Loop for all of the lines of text
        for line in text:

            # 2. Create a hand
            a_hand = Hand(text=line, part2=self.part2)

            # 3. And save it
            self.hands.append(a_hand)

    def winnings(self):
        "Determing the total winnings for the hands"

        # 1. Start with nothing
        result = 0

        # 2. Sort the hands
        sorted_hands = sorted(self.hands, key=lambda h: h.key)

        # 3. Collect the winnings hand by hand
        for rank, one_hand in enumerate(sorted_hands):

            # 4. Add in the winnings for this hand
            result += one_hand.winnings(rank + 1)

        # 5. Return the total winnings
        return result


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------

if __name__ == '__main__':
    pass

# ======================================================================
# end                      h a n d s . p y                     end
# ======================================================================

File: 2023/07_CamelCards/hands.go
// ======================================================================
// Camel Cards
//   Advent of Code 2023 Day 07 -- Eric Wastl -- https://adventofcode.com
//
// go implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                         h a n d s . g o
// ======================================================================
// A solver for the Advent of Code 2023 Day 07 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"fmt"
	"sort"
)

// ----------------------------------------------------------------------
//
//	constants
//
// ----------------------------------------------------------------------
const ()

// ======================================================================
//                                                                  Hands
// ======================================================================

// Hands ... Object for Camel Cards
type Hands struct {
	Part2 bool
	Text  []string
	Hands []*Hand
}

// NewHands ... Create a Hands object
func NewHands(part2 bool, text []string) (*Hands, error) {

	// 1. Set the initial values
	p := new(Hands)
	p.Part2 = part2
	p.Text = text

	// 2. Process text (if any)
	if len(p.Text) > 0 {
		err := p.processText(p.Text)
		if err != nil {
			return nil, err
		}
	}

	// 3. Return the new Hands object
	return p, nil
}

// processText ... Assign values from text
func (p *Hands) processText(text []string) error {

	// 1. Loop for each line of the text
	for indx, line := range text {

		// 2. Create a hand
		hand, err := NewHand(p.Part2, line)
		if err != nil {
			return fmt.Errorf("text line %d: %s", indx, err)
		}

		// 3. Add it in
		p.Hands = append(p.Hands, hand)
	}

	// 4. Return success
	return nil
}

// Sort ... Soft the hands (worst is 0)
func (p *Hands) Sort() {

	// 1. Sort the hands
	sort.Slice(p.Hands, func(i, j int) bool {
		return p.Hands[i].Less(p.Hands[j])
	})
}

// PartOne ... Returns the solution for part one
func (p *Hands) PartOne(verbose bool, limit int) string {

	// 1. Sort the hands
	p.Sort()

	// 2. Start with nothing
	winnings := 0

	// 3. Loop for all the hands
	for indx, hand := range p.Hands {

		// 4. Add to the winnings
		winnings = winnings + (indx+1)*hand.Bid
	}

	// 5. Return the solution for part one
	return fmt.Sprint(winnings)
}

// PartTwo ... Returns the solution for part two
func (p *Hands) PartTwo(verbose bool, limit int) string {

	// 1. Sort the hands
	p.Sort()

	// 2. Start with nothing
	winnings := 0

	// 3. Loop for all the hands
	for indx, hand := range p.Hands {

		// 4. Add to the winnings
		winnings = winnings + (indx+1)*hand.Bid
	}

	// 5. Return the solution for part two
	return fmt.Sprint(winnings)
}

// ======================================================================
// end                         h a n d s . g o                        end
// ======================================================================
